substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -2160,164 +2169,102 @@ public NativeImageDebugLocationInfo split(int stackDecrement, int adjustment) {
+                // unadjusted stack pointer is just the raw offset
+                adjustment = AARCH64_PRE_EXTEND_FRAME_ADJUSTMENT;
             }
         }
 
-        public NativeImageDebugLocalValue nextFloatingLocation() {
-            if (nextFPregIdx < fregs.length) {
-                return NativeImageDebugRegisterValue.create(fregs[nextFPregIdx++].number);
-            } else {
-                return nextStackLocation();
-            }
+        ParamLocationProducer(HostedMethod method) {
+            this.hostedMethod = method;
+            this.callingConvention = getCallingConvention(hostedMethod);
+            // assume no stack slots until we find out otherwise
+            this.usesStack = false;
         }
 
-        public NativeImageDebugLocalValue nextIntegerLocation() {
-            if (nextGPRegIdx < gpregs.length) {
-                return NativeImageDebugRegisterValue.create(gpregs[nextGPRegIdx++].number);
-            } else {
-                return nextStackLocation();
-            }
+        NativeImageDebugLocalValue thisLocation() {
+            assert !hostedMethod.isStatic();
+            return unpack(callingConvention.getArgument(0));
         }
 
-        public NativeImageDebugLocalValue nextStackLocation() {
-            // offset is computed relative to the undecremented stack pointer and includes an extra
-            // offset to adjust for any intervening return address and frame pointer
-            assert nextStackIdx < stackParamCount : "encountered too many stack params";
-            int stackIdx = nextStackIdx++;
-            return NativeImageDebugStackValue.create((stackIdx * 8) + stackOffset);
+        NativeImageDebugLocalValue paramLocation(int paramIdx) {
+            assert paramIdx < hostedMethod.getSignature().getParameterCount(false);
+            int idx = paramIdx;
+            if (!hostedMethod.isStatic()) {
+                idx++;
+            }
+            return unpack(callingConvention.getArgument(idx));
         }
 
-        public boolean usesStack() {
-            return stackParamCount > 0;
+        private NativeImageDebugLocalValue unpack(AllocatableValue value) {
+            if (value instanceof RegisterValue) {
+                RegisterValue registerValue = (RegisterValue) value;
+                return NativeImageDebugRegisterValue.create(registerValue);
+            } else {
+                // call argument must be a stack slot if it is not a register
+                StackSlot stackSlot = (StackSlot) value;
+                this.usesStack = true;
+                // the calling convention provides raw offsets from the SP at point
+                // of call but these need adjusting to allow for a stacked return
+                // address and/or frame pointer according to the architecture
+                return new NativeImageDebugStackValue(stackSlot.getRawOffset() + adjustment);
```

Comment:
Shouldn't we be using `create` here?
```suggestion
                return NativeImageDebugStackValue.create(stackSlot, adjustment);
```

Also, there's no need to use `getRawOffset` because `StackSlot.getOffset` calculates the correct location of the incoming stack arguments as long as you use the correct frame size at the point of interest. And since at the method entry point we only have the part of the frame that contains the return address, the size of the frame is equal to the size of the return address, which is what `adjustment` actually is.

Reply:
Yes, we can use `memoizedCreate(int)` to avoid duplication of the same value.
I renamed `PRE_EXTEND_FRAME_ADJUSTMENT` to `PRE_EXTEND_FRAME_SIZE` and passed `getOffset(PRE_EXTEND_FRAME_SIZE)` to `memoizedCreate`.

Reply:
@pejovica The change suggested above results in a failure on AArch64. The call to 
```
NativeImageDebugStackValue.create(stackSlot, PRE_EXTEND_FRAME_SIZE)
```
results in a call to
```
NativeImageDebugStackValue.memoizedCreate(value.getOffset(PRE_EXTEND_FRAME_SIZE))
```
On AARCH64 we have `PRE_EXTEND_FRAME_SIZE == 0` which means that we run up against the following assert in `StackSlot.getOffset()`
```
public int getOffset(int totalFrameSize) {
    assert totalFrameSize > 0 || !addFrameSize;
    . . .
```
The problem is that the pre-frame push stack slots coming out of the AArch64 calling convention satisfy `addFrameSize == true` even though at this point on AArch64 the total frame size is zero. The problem does not arise on x86_64 because the call will have pushed a return address meaning that the total frame size will not be zero.

I have finessed this in NativeImageDebugStackValue.create by special casing `framesize == 0` to call `StackSlot.getRawOffset()`.

Reply:
Looks good, thanks!